netutils/acl.py

```diff
@@ -0,0 +1,354 @@
+from netutils.protocol_mapper import PROTO_NAME_TO_NUM, TCP_NAME_TO_NUM, UDP_NAME_TO_NUM
+from netutils.ip import is_ip_within
+
+try:
+    import jsonschema
+
+    HAS_JSON_SCHEMA = True
+except ImportError:
+    HAS_JSON_SCHEMA = False
+
+DATA_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "name": {"type": "string"},
+        "src_zone": {"type": ["string", "array"]},
+        "src_ip": {"$ref": "#/definitions/arrayOrIP"},
+        "dst_ip": {"$ref": "#/definitions/arrayOrIP"},
+        "dst_port": {
+            "oneOf": [
+                {
+                    "$ref": "#/definitions/port",
+                },
+                {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/port",
+                    },
+                    "minItems": 1,
+                    "uniqueItems": True,
+                },
+            ],
+        },
+        "dst_zone": {"type": ["string", "array"]},
+        "action": {"type": "string"},
+    },
+    "definitions": {
+        "ipv4": {"type": "string", "pattern": "^(?:\\d{1,3}\\.){3}\\d{1,3}$"},
+        "ipv4_cidr": {"type": "string", "pattern": "^(?:\\d{1,3}\\.){3}\\d{1,3}\\/\\d{1,2}$"},
+        "ipv4_range": {"type": "string", "pattern": "^(?:\\d{1,3}\\.){3}\\d{1,3}\\-(?:\\d{1,3}\\.){3}\\d{1,3}$"},
+        "ipv6": {"type": "string", "pattern": "^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$"},
+        "ipv6_cidr": {"type": "string", "pattern": "^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}/[0-9]{1,3}$"},
+        "ipv6_range": {
+            "type": "string",
+            "pattern": "^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}-([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$",
+        },
+        "oneIP": {
+            "oneOf": [
+                {
+                    "$ref": "#/definitions/ipv4",
+                },
+                {
+                    "$ref": "#/definitions/ipv4_cidr",
+                },
+                {
+                    "$ref": "#/definitions/ipv4_range",
+                },
+                {
+                    "$ref": "#/definitions/ipv6",
+                },
+                {
+                    "$ref": "#/definitions/ipv6_cidr",
+                },
+                {
+                    "$ref": "#/definitions/ipv6_range",
+                },
+            ],
+        },
+        "arrayOrIP": {
+            "oneOf": [
+                {
+                    "$ref": "#/definitions/oneIP",
+                },
+                {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/oneIP",
+                    },
+                },
+            ],
+        },
+        "port": {"type": "string", "pattern": "^\\S+\\/\\S+$"},
+    },
+    "required": [],
+}
+
+
+AS_DICT_SCHEMA = copy.deepcopy(DATA_SCHEMA)
+AS_DICT_SCHEMA["definitions"]["port"]["pattern"] = "^\\d+\\/\\d+$"
+
+
+def _cartesian_product(dictionary: t.Dict[str, str]):
```

Comment:
a docstring with an example would help understanding what this function is returning